akka-http2-support/src/main/scala/akka/http/impl/engine/http2/RequestParsing.scala

```diff
@@ -35,72 +39,112 @@ private[http2] object RequestParsing {
       ): HttpRequest =
         if (remainingHeaders.isEmpty) {
           // 8.1.2.3: these pseudo header fields are mandatory for a request
-          checkRequiredField(":scheme", scheme)
-          checkRequiredField(":method", method)
-          checkRequiredField(":path", path)
+          checkRequiredPseudoHeader(":scheme", scheme)
+          checkRequiredPseudoHeader(":method", method)
+          checkRequiredPseudoHeader(":path", pathAndRawQuery)
 
           headers += Http2StreamIdHeader(subStream.streamId)
+          if (cookies != null) {
+            // Compress 'cookie' headers if present
+            headers += parseHeaderPair(httpHeaderParser, "cookie", cookies.toString)
+          }
           if (remoteAddressHeader.isDefined) headers += remoteAddressHeader.get
 
           val entity =
             if (subStream.data == Source.empty || contentLength == 0) HttpEntity.Empty
             else if (contentLength > 0) HttpEntity.Default(contentType, contentLength, subStream.data)
             else HttpEntity.Chunked.fromData(contentType, subStream.data)
 
-          val uri = Uri(scheme, authority).withPath(Uri.Path(path))
+          val (path, rawQueryString) = pathAndRawQuery
+          val uri = Uri(scheme, authority, path, rawQueryString)
           HttpRequest(
             method, uri, headers.result(), entity, HttpProtocols.`HTTP/2.0`
           )
         } else remainingHeaders.head match {
           case (":scheme", value) ⇒
-            rec(remainingHeaders.tail, method, value, authority, path, contentType, contentLength, headers)
+            checkUniquePseudoHeader(":scheme", scheme)
+            checkNoRegularHeadersBeforePseudoHeader(":scheme", seenRegularHeader)
+            rec(remainingHeaders.tail, method, value, authority, pathAndRawQuery, contentType, contentLength, cookies, seenRegularHeader, headers)
           case (":method", value) ⇒
+            checkUniquePseudoHeader(":method", method)
+            checkNoRegularHeadersBeforePseudoHeader(":method", seenRegularHeader)
             val m = HttpMethods.getForKey(value).getOrElse(malformedRequest(s"Unknown HTTP method: '$value'"))
-            rec(remainingHeaders.tail, m, scheme, authority, path, contentType, contentLength, headers)
-          case (":path", path) ⇒
-            rec(remainingHeaders.tail, method, scheme, authority, path, contentType, contentLength, headers)
+            rec(remainingHeaders.tail, m, scheme, authority, pathAndRawQuery, contentType, contentLength, cookies, seenRegularHeader, headers)
+          case (":path", value) ⇒
+            checkUniquePseudoHeader(":path", pathAndRawQuery)
+            checkNoRegularHeadersBeforePseudoHeader(":path", seenRegularHeader)
+            // QUESTION: Should this logic be exposed in a method in either Uri or Query? (would be more consistent)
+            // QUESTION: Should we catch the parsing exception and throw a more specific type?
```

Comment:
Yes, let's do it like in the HTTP/1 parser here: https://github.com/akka/akka-http/blob/3c966af2255b58b02fbdacd6abec89c47f78ce48/akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpRequestParser.scala#L141-L146

Reply:
OK, I'll add a method to `Uri`, call it from `RequestParsing` and wrap the call from `RequestParsing` in a `try`/`catch` block.